shell/common/animator.cc

```diff
@@ -95,9 +106,27 @@ void Animator::BeginFrame(fml::TimePoint frame_start_time,
   }
 
   if (!frame_scheduled_) {
-    // We don't have another frame pending, so we're waiting on user input
-    // or I/O. Allow the Dart VM 100 ms.
-    delegate_.OnAnimatorNotifyIdle(*this, dart_frame_deadline_ + 100000);
+    // Under certain workloads (such as our parent view resizing us, which is
+    // communicated to us by repeat viewport metrics events), we won't
+    // actually have a frame scheduled yet, despite the fact that we *will* be
+    // producing a frame next vsync (it will be scheduled once we receive the
+    // viewport event).  Because of this, we hold off on calling
+    // |OnAnimatorNotifyIdle| for a little bit, as that could cause garbage
+    // collection to trigger at a highly undesirable time.
+    task_runners_.GetUITaskRunner()->PostDelayedTask(
+        [self = weak_factory_.GetWeakPtr(),
+         notify_idle_task_id = notify_idle_task_id_]() {
+          if (!self.get()) {
+            return;
+          }
+          // If our (this task's) task id is the same as the current one, then
+          // no further frames were produced, and it is safe (w.r.t. jank) to
+          // notify the engine we are idle.
+          if (notify_idle_task_id == self->notify_idle_task_id_) {
+            self->delegate_.OnAnimatorNotifyIdle(
+                *self.get(), self->dart_frame_deadline_ + 100000);
```

Comment:
Since this runs significantly later now, can we change this to Dart_TimelineGetMicros() + 100000? (I.e., allow for 100ms after the delay instead of 100ms from the original frame.)

Reply:
Ah, good call.  Done.